File: media-manager/backend/app/schemas/download.py
"""Download schemas for API validation."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, HttpUrl


class DownloadCreate(BaseModel):
    """Schema for creating a new download."""

    url: str = Field(..., description="URL to download")
    premium_account_id: Optional[int] = Field(None, description="Premium account ID to use")


class DownloadBulkCreate(BaseModel):
    """Schema for creating multiple downloads."""

    urls: list[str] = Field(..., description="List of URLs to download")
    premium_account_id: Optional[int] = Field(None, description="Premium account ID to use")


class DownloadResponse(BaseModel):
    """Schema for download response."""

    id: int
    url: str
    filename: Optional[str] = None
    status: str
    progress: float
    speed: float
    size_total: Optional[int] = None
    size_downloaded: int
    eta: Optional[int] = None
    aria2_gid: Optional[str] = None
    premium_account_id: Optional[int] = None
    file_path: Optional[str] = None
    error_message: Optional[str] = None
    retry_count: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DownloadUpdate(BaseModel):
    """Schema for updating a download."""

    status: Optional[str] = None
    filename: Optional[str] = None
    file_path: Optional[str] = None


class DownloadStats(BaseModel):
    """Schema for global download statistics."""

    total_downloads: int
    completed_downloads: int
    failed_downloads: int
    active_downloads: int
    current_speed: float
    num_active: int
    num_waiting: int


class DownloadActionResponse(BaseModel):
    """Schema for download action responses."""

    success: bool
    message: str
    download: Optional[DownloadResponse] = None

File: media-manager/frontend/src/types/download.ts
export interface Download {
  id: number;
  url: string;
  filename: string | null;
  status: DownloadStatus;
  progress: number;
  speed: number;
  size_total: number | null;
  size_downloaded: number;
  eta: number | null;
  aria2_gid: string | null;
  premium_account_id: number | null;
  file_path: string | null;
  error_message: string | null;
  retry_count: number;
  created_at: string;
  updated_at: string | null;
}

export type DownloadStatus =
  | 'pending'
  | 'queued'
  | 'downloading'
  | 'completed'
  | 'failed'
  | 'paused'
  | 'cancelled';

export interface DownloadStats {
  total_downloads: number;
  completed_downloads: number;
  failed_downloads: number;
  active_downloads: number;
  current_speed: number;
  num_active: number;
  num_waiting: number;
}

export interface CreateDownloadRequest {
  url: string;
  premium_account_id?: number;
}

export interface BulkDownloadRequest {
  urls: string[];
  premium_account_id?: number;
}

export interface DownloadActionResponse {
  success: boolean;
  message: string;
  download?: Download;
}
